### ISSUE TITLE
VSO can't create Secret due to permission denied

### ISSUE THREAD

**Issue by mochizuki875** [hashicorp/vault-secrets-operator#207]
**Describe the bug**
After setting a new secret and binding the policy to a role, VSO fails to create the `Secret` correspond to `VaultDynamicSecret`.

**To Reproduce**

It is a bit complicated.

1. Deploy Vault and VSO via Helm charts
2. Set up kubernetes auth
3. Set up KV secrets engine
4. Bind policy for KV secrets engine to role
5. Create `VaultConnection` and `VaultAuth`.
6. Create `VaultStaticSecret`(It will work well)
7. Set up Database secrets engine
8. Bind policy for Database secrets engine to the same role
9. Create `VaultDynamicSecret`(It will not work well)

#### Preparation

Deploy Vault and VSO via Helm charts.
In this case, they are installed in vault namespace.
So FQDN to Vault is `http://vault.vault.svc:8200`.

```console
$ kubectl get po -n vault
NAME                                                         READY   STATUS    RESTARTS   AGE
vault-0                                                      1/1     Running   0          112s
vault-agent-injector-57db6b66cf-k46pb                        1/1     Running   0          113s
vault-csi-provider-c7sv5                                     2/2     Running   0          113s
vault-secrets-operator-controller-manager-67cfc56bcd-l7t74   2/2     Running   0          101s

$ kubectl get svc -owide -n vault
NAME                                     TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)             AGE     SELECTOR
vault                                    ClusterIP   10.103.40.100    <none>        8200/TCP,8201/TCP   2m16s   app.kubernetes.io/instance=vault,app.kubernetes.io/name=vault,component=server
vault-agent-injector-svc                 ClusterIP   10.101.199.185   <none>        443/TCP             2m16s   app.kubernetes.io/instance=vault,app.kubernetes.io/name=vault-agent-injector,component=webhook
vault-internal                           ClusterIP   None             <none>        8200/TCP,8201/TCP   2m16s   app.kubernetes.io/instance=vault,app.kubernetes.io/name=vault,component=server
vault-secrets-operator-metrics-service   ClusterIP   10.104.115.51    <none>        8443/TCP            2m4s    app.kubernetes.io/instance=vault-secrets-operator,app.kubernetes.io/name=vault-secrets-operator,control-plane=controller-manager
```

Enable Kubernetes auth.
```console
/ $ vault auth enable kubernetes
Success! Enabled kubernetes auth method at: kubernetes/
/ $ 
/ $ vault write auth/kubernetes/config \
>     kubernetes_host=https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT
Success! Data written to: auth/kubernetes/config
```

Create `Service Account`.
```console
$ kubectl create sa app
serviceaccount/app created
```

#### Step1: KV Secret

Enable KV secrets engine and populate values.
```console
/ $ vault secrets enable -path=secret kv-v2
Success! Enabled the kv-v2 secrets engine at: secret/

/ $ vault kv put secret/app/config USERNAME="user12345" PASSWORD="pass12345"
===== Secret Path =====
secret/data/app/config

======= Metadata =======
Key                Value
---                -----
created_time       2023-05-24T14:12:50.300151539Z
custom_metadata    <nil>
deletion_time      n/a
destroyed          false
version            1
```

Create policy for KV secrets engine.
```console
/ $ vault policy write app-secret - <<EOF
> path "secret/data/app/config" {
>  capabilities = ["read"]
> }
> EOF
Success! Uploaded policy: app-secret
```

Create role and bind the policy.
```console
/ $ vault write auth/kubernetes/role/app \
> bound_service_account_names=app \
> bound_service_account_namespaces=default \
> policies=app-secret \
> ttl=24h
Success! Data written to: auth/kubernetes/role/app
```

Create `VaultConnection` and  `VaultAuth`.
```console
$ kubectl apply -f app-vault-connection.yaml
vaultconnection.secrets.hashicorp.com/app-vault-connection created

$ kubectl apply -f app-vault-auth.yaml
vaultauth.secrets.hashicorp.com/app-vault-auth created
```

<details>

`app-vault-connection.yaml`
```yaml
apiVersion: secrets.hashicorp.com/v1alpha1
kind: VaultConnection
metadata:
  name: app-vault-connection
  namespace: default
spec:
  address: http://vault.vault.svc:8200
```

`app-vault-auth.yaml`
```yaml
apiVersion: secrets.hashicorp.com/v1alpha1
kind: VaultAuth
metadata:
  name: app-vault-auth
  namespace: default
spec:
  vaultConnectionRef: app-vault-connection
  method: kubernetes
  mount: kubernetes
  kubernetes:
    role: app
    serviceAccount: app
```

</details>

Create `VaultStaticSecret` and check `Secret`.
It works well.
```console
$ kubectl apply -f app-vault-static-secret.yaml
vaultstaticsecret.secrets.hashicorp.com/app-vault-static-secret created

$ kubectl get VaultStaticSecret
NAME                      AGE
app-vault-static-secret   25s

$ kubectl get secret
NAME                  TYPE     DATA   AGE
app-secret-from-vso   Opaque   3      18s
```

<details>

`app-vault-static-secret.yaml`
```yaml
apiVersion: secrets.hashicorp.com/v1alpha1
kind: VaultStaticSecret
metadata:
  name: app-vault-static-secret
  namespace: default
spec:
  vaultAuthRef: app-vault-auth
  type: kv-v2
  mount: secret
  name: app/config
  refreshAfter: 60s
  destination:
    create: true
    name: app-secret-from-vso
```

</details>

#### Step2: Database Secret

First, deploy mysql pod.
```console
$ kubectl apply -f mysql.yaml 
secret/mysql-secret created
deployment.apps/mysql created
service/mysql created
```

<details>

`mysql.yaml`
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: mysql-secret
type: kubernetes.io/basic-auth
stringData:
  password: rootpass
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mysql
spec:
  selector:
    matchLabels:
      app: mysql
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        app: mysql
    spec:
      containers:
      - image: mysql:8.0.33
        name: mysql
        env:
        - name: MYSQL_ROOT_PASSWORD
          valueFrom:
            secretKeyRef:
              name: mysql-secret
              key: password
        ports:
        - containerPort: 3306
          name: mysql
---
apiVersion: v1
kind: Service
metadata:
  name: mysql
spec:
  ports:
  - port: 3306
  selector:
    app: mysql
```

</details>

Enter mysql and create database.
```console
$ kubectl exec -it $(kubectl get pods --selector "app=mysql" -o jsonpath="{.items[0].metadata.name}") -c mysql -- bash -c 'mysql -u root -p'
Enter password: rootpass
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 8
Server version: 8.0.33 MySQL Community Server - GPL

Copyright (c) 2000, 2023, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> create database workshop;
Query OK, 1 row affected (0.00 sec)

mysql> exit
Bye
```

Enable and configure Database secrets engine.
```console
/ $ vault secrets enable database
Success! Enabled the database secrets engine at: database/

/ $ vault write database/config/workshop-db \
> plugin_name=mysql-database-plugin \
> connection_url="{{username}}:{{password}}@tcp(mysql.default.svc.cluster.local:3306)/" \
> allowed_roles="workshop-role" \
> username="root" \
> password="rootpass"
Success! Data written to: database/config/workshop-db
```

Create role.
```console
/ $ vault write database/roles/workshop-role \
> db_name=workshop-db \
> creation_statements="CREATE USER '{{name}}'@'%' IDENTIFIED BY '{{password}}';GRANT SELECT ON *.* TO '{{name}}'@'%';" \
> default_ttl="1h" \
> max_ttl="24h"
Success! Data written to: database/roles/workshop-role
```

Then, Dynamic Secret can be generated as below.
```console
/ $ vault read database/creds/workshop-role
Key                Value
---                -----
lease_id           database/creds/workshop-role/saqCENZ4db2Qgp7srJmlnDag
lease_duration     1h
lease_renewable    true
password           9KSZo-TDpCFcVcWsA4TT
username           v-root-workshop-r-uo82kHqDC2eq0w
```

Create policy.
```console
/ $ vault policy write db-app - <<EOF
> path "database/creds/workshop-role" {
>   capabilities = ["read"]
> }
> EOF
Success! Uploaded policy: db-app
```

Bind policy to role for kubernetes auth (created step1).
```
/ $ vault write auth/kubernetes/role/app \
> bound_service_account_names=app \
> bound_service_account_namespaces=default \
> policies=app-secret,db-app \
> ttl=24h
Success! Data written to: auth/kubernetes/role/app
```

Create `VaultDynamicSecret` and check `Secret`.
It dosen't works.
`Secret` correspond to `VaultDynamicSecret` will not be created.
```console
$ kubectl apply -f app-vault-dynamic-secret.yaml
vaultdynamicsecret.secrets.hashicorp.com/mysql-dynamic-secret created

$ kubectl get VaultDynamicSecret
NAME                   AGE
mysql-dynamic-secret   12s

$ kubectl get secret
NAME                  TYPE                       DATA   AGE
app-secret-from-vso   Opaque                     3      15m
mysql-secret          kubernetes.io/basic-auth   1      12m
```

<details>

`app-vault-dynamic-secret.yaml`
```yaml
apiVersion: secrets.hashicorp.com/v1alpha1
kind: VaultDynamicSecret
metadata:
  name: mysql-dynamic-secret
  namespace: default
spec:
  vaultAuthRef: app-vault-auth
  role : workshop-role
  mount: database
  destination:
    create: true
    name: mysql-secret-from-vso
```

</details>

VSO log is as below.
```console
$ $ kubectl -n vault logs vault-secrets-operator-controller-manager-67cfc56bcd-l7t74
...
2023-05-24T15:05:34Z	ERROR	Reconciler error	{"controller": "vaultdynamicsecret", "controllerGroup": "secrets.hashicorp.com", "controllerKind": "VaultDynamicSecret", "VaultDynamicSecret": {"name":"mysql-dynamic-secret","namespace":"default"}, "namespace": "default", "name": "mysql-dynamic-secret", "reconcileID": "86bb53c3-d7b6-4894-8083-7d1d377ee23a", "error": "Error making API request.\n\nURL: GET http://vault.vault.svc:8200/v1/database/creds/workshop-role\nCode: 403. Errors:\n\n* 1 error occurred:\n\t* permission denied\n\n"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:329
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:235
...
```

If we recreate `VaultAuth`, it will work well.
```console
$ kubectl delete -f app-vault-dynamic-secret.yaml
vaultdynamicsecret.secrets.hashicorp.com "mysql-dynamic-secret" deleted

$ kubectl delete -f app-vault-auth.yaml
vaultauth.secrets.hashicorp.com "app-vault-auth" deleted

$ kubectl apply -f app-vault-auth.yaml
vaultauth.secrets.hashicorp.com/app-vault-auth created

$ kubectl apply -f app-vault-dynamic-secret.yaml
vaultdynamicsecret.secrets.hashicorp.com/mysql-dynamic-secret created

$ kubectl get secret
NAME                    TYPE                       DATA   AGE
app-secret-from-vso     Opaque                     3      8m1s
mysql-secret            kubernetes.io/basic-auth   1      5m45s
mysql-secret-from-vso   Opaque                     3      8s
```

#### Appendix: Another KV Secret

In case of another KV secret, it will not work well too.
It is reported #206.

Enable another KV secrets engine and populate values.
```console
/ $ vault secrets enable -path=another-secret kv-v2
Success! Enabled the kv-v2 secrets engine at: another-secret/

/ $ vault kv put another-secret/app/config USERNAME="USER12345" PASSWORD="PASS12345"
========= Secret Path =========
another-secret/data/app/config

======= Metadata =======
Key                Value
---                -----
created_time       2023-05-24T15:33:57.944273805Z
custom_metadata    <nil>
deletion_time      n/a
destroyed          false
version            1
```

Create policy for another KV secrets engine.
```console
/ $ vault policy write app-another-secret - <<EOF
> path "another-secret/data/app/config" {
>  capabilities = ["read"]
> }
> EOF
Success! Uploaded policy: app-another-secret
```

Create role and bind the policy.
```console
/ $ vault write auth/kubernetes/role/app \
> bound_service_account_names=app \
> bound_service_account_namespaces=default \
> policies=app-secret,db-app,app-another-secret \
> ttl=24h
Success! Data written to: auth/kubernetes/role/app
```

Create `VaultStaticSecret` and the `Secret` correspond to it is not created.
```console
$ kubectl apply -f app-vault-static-another-secret.yaml
vaultstaticsecret.secrets.hashicorp.com/app-vault-another-static-secret created

$ kubectl get VaultStaticSecret
NAME                              AGE
app-vault-another-static-secret   23s
app-vault-static-secret           39m

$ kubectl get secret
NAME                    TYPE                       DATA   AGE
app-secret-from-vso     Opaque                     3      39m
mysql-secret            kubernetes.io/basic-auth   1      37m
mysql-secret-from-vso   Opaque                     3      31m
```

<details>

`app-vault-static-another-secret.yaml`
```yaml
---
apiVersion: secrets.hashicorp.com/v1alpha1
kind: VaultStaticSecret
metadata:
  name: app-vault-another-static-secret
  namespace: default
spec:
  vaultAuthRef: app-vault-auth
  type: kv-v2
  mount: another-secret
  name: app/config
  refreshAfter: 60s
  destination:
    create: true
    name: app-another-secret-from-vso
```

</details>

The VSO log is as below.
```console
$ kubectl -n vault logs vault-secrets-operator-controller-manager-67cfc56bcd-l7t74
...
2023-05-24T15:38:53Z	ERROR	empty Vault secret	{"controller": "vaultstaticsecret", "controllerGroup": "secrets.hashicorp.com", "controllerKind": "VaultStaticSecret", "VaultStaticSecret": {"name":"app-vault-another-static-secret","namespace":"default"}, "namespace": "default", "name": "app-vault-another-static-secret", "reconcileID": "805e7430-302c-49c5-9a80-de679f682d28", "mount": "another-secret", "name": "app/config"}
github.com/hashicorp/vault-secrets-operator/controllers.(*VaultStaticSecretReconciler).Reconcile
	/home/runner/work/vault-secrets-operator/vault-secrets-operator/controllers/vaultstaticsecret_controller.go:112
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:122
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:323
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:235
...
```

**Expected behavior**
The `Secret` correspond to `VaultDynamicSecret` is created.

**Environment**
* Kubernetes version: 
```console
$ minikube version
minikube version: v1.30.1
commit: 08896fd1dc362c097c925146c4a0d0dac715ace0

$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.1", GitCommit:"4c9411232e10168d7b050c49a1b59f6df9d7ea4b", GitTreeState:"clean", BuildDate:"2023-04-14T13:21:19Z", GoVersion:"go1.20.3", Compiler:"gc", Platform:"linux/amd64"}
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.1", GitCommit:"4c9411232e10168d7b050c49a1b59f6df9d7ea4b", GitTreeState:"clean", BuildDate:"2023-04-14T13:21:19Z", GoVersion:"go1.20.3", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
```

* vault-secrets-operator and vault version:
```console
$ helm list -n vault
NAME                  	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                            	APP VERSION
vault                 	vault    	1       	2023-05-24 23:55:14.988542768 +0900 JST	deployed	vault-0.24.1                     	1.13.1     
vault-secrets-operator	vault    	1       	2023-05-24 23:55:26.987889309 +0900 JST	deployed	vault-secrets-operator-0.1.0-beta	0.1.0-beta 
```

**Comment by benashz** [comment]
Thank you for the detailed issue report @mochizuki875 ! We currently looking into this issue.

**Comment by teksuo** [comment]
Same here with some simple VaultStaticSecret. Policies were attached to the role after everything else was setup.

Only solution I found was to uninstall and install VSO again, then the secrets were correctly synced.

**Comment by mochizuki875** [comment]
Hi, I've checked it in recently [GA](https://www.hashicorp.com/blog/vault-secrets-operator-for-kubernetes-now-ga) release version v0.1.0.
It seems to be fixed v0.1.0!

**Comment by benashz** [comment]
Fix should be included in https://github.com/hashicorp/vault-secrets-operator/releases/tag/v0.2.0
